rfd/0234-in-band-mfa-ssh-sessions.md

```diff
@@ -0,0 +1,756 @@
+1. MFA complexity is moved from clients to the SSH service, so clients only need to connect to the SSH service and
+   resolve an MFA challenge only if required, streamlining the session establishment process and reducing the risk of
+   implementation errors.
+1. Per-session MFA certificates can be completely removed, eliminating a single credential representing multiple factors
+   of authentication.
+
+In summary, this RFD proposes a more secure and streamlined approach to MFA enforcement for SSH sessions by integrating
+MFA checks directly into the session establishment process, reducing client complexity, and eliminating the need for
+per-session MFA certificates.
+
+## Non-Goals
+
+1. This RFD does not propose changes to other Teleport access protocols such as Kubernetes, databases, desktops, etc.
+   However, the architecture could be extended to these protocols in the future.
+
+## Details
+
+### UX
+
+No changes are expected since this is an internal change.
+
+### High-Level Flow
+
+The client will first dial its target SSH host using the Proxy and its standard Teleport client certificate. The Proxy
+will then resolve the target cluster and host, and invoke the `EvaluateSSHAccess` RPC of the [Decision
+service](https://github.com/gravitational/Teleport.e/blob/master/rfd/0024e-access-control-decision-api.md). Upon
+receiving a permit from the Decision service, the Proxy will staple the permit to the connection and open a connection
+to the respective cluster's SSH service.
+
+The SSH service will authenticate the client using the provided client certificate and the stapled permit. The SSH
+service will then check if MFA is required for the session by examining the permit.
+
+If MFA is _not required_, the SSH service will then proceed to establish the SSH session.
+
+If MFA _is required_, the SSH service will send a Protobuf [`MFAPrompt`
+message](#ssh-keyboard-interactive-authentication) as a Question via the SSH [keyboard-interactive
+authentication](https://www.rfc-editor.org/rfc/rfc4256) method to inform the client that MFA is needed. The client must
+then call the `CreateChallenge` RPC on the MFA service, providing a _Session Identifying Payload (SIP)_ and any relevant
+metadata. The SIP is an SSH session hash computed from SSH session state and is used to bind the MFA challenge to the
+specific session. Both the SSH client and the SSH server can independently compute the SIP from session state.
+
+The MFA service will [store the Session Identifying Payload (SIP)](#storing-session-identifying-payloads) and respond
+with an MFA challenge for the client to solve.
+
+Next, the client solves the MFA challenge and calls `ValidateChallenge` with the MFA response. The MFA service validates
+the MFA response and stores the `ValidatedChallenge` to the local backend.
+
+In addition, if the target cluster is a leaf cluster, a watcher in the reverse tunnel server monitors
```

Comment:
If we returned some sort of token to the client on successful validation, say a JWT, the SSH service wouldn't need watchers or replication. I think it's an alternative worth mentioning.

Reply:
The current state of affairs is that we return a cryptographic proof of MFA validation in the form of a SSH certificate and the entire reason for this endeavor comes from the fact that we botched that validation badly, why would we have better luck with a JWT?